### PR TITLE
Added weight tying

### PR DESCRIPTION
**Pull request by ofirpress**
As we recently showed in [1], tying the input word embedding to the softmax matrix decreases perplexity on a wide range of language models.

I implemented weight tying, and also ran tests comparing weight tied models to "regular" models, to make sure that it also improves the performance of the various models in this repo. The results are:

> 
>REGULAR:
>-- rnn_cell = rnn_cell_modern.MGUCell(size, use_multiplicative_integration = True, use_recurrent_dropout = False)
>Epoch: 13 Train Perplexity: 50.346
>Epoch: 13 Valid Perplexity: 131.284
>Test Perplexity: 125.133
>
>-- rnn_cell = rnn_cell_mulint_layernorm_modern.HighwayRNNCell_MulInt_LayerNorm(size)
>Epoch: 13 Train Perplexity: 54.869
>Epoch: 13 Valid Perplexity: 123.373
>Test Perplexity: 117.859
>
>-- rnn_cell = rnn_cell_mulint_modern.GRUCell_MulInt(size)
>Epoch: 13 Train Perplexity: 49.530
>Epoch: 13 Valid Perplexity: 126.091
>Test Perplexity: 120.613

>WEIGHT TIED:
>-- rnn_cell = rnn_cell_modern.MGUCell(size, use_multiplicative_integration = True, use_recurrent_dropout = False)
>Epoch: 13 Train Perplexity: 61.479
>Epoch: 13 Valid Perplexity: 128.844
>Test Perplexity: 122.353

>-- rnn_cell = rnn_cell_mulint_layernorm_modern.HighwayRNNCell_MulInt_LayerNorm(size)
>Epoch: 13 Train Perplexity: 64.139
>Epoch: 13 Valid Perplexity: 120.395
>Test Perplexity: 114.491

>-- rnn_cell = rnn_cell_mulint_modern.GRUCell_MulInt(size)
>Epoch: 13 Train Perplexity: 60.983
>Epoch: 13 Valid Perplexity: 125.205
>Test Perplexity: 119.398

Clearly weight tying consistently reduces test set perplexity while increasing the train perplexity which indicates less overfitting. 

[1] https://arxiv.org/abs/1608.05859

### CODE REVIEW

**Comment by LeavesBreathe** [comment]
Thanks for implementing these changes. This will reduce on memory usage as well and decrease parameters.
